### PR TITLE
error_model_alignment

### LINKED ISSUE #127: Which error send back when issue with Device Identifier provided by the API Consumer?
**Problem description**
A very small issue but looking for feedback. I guess we can have a common behavior.

In a lot of our API we use the device identifier schema to get the identifier of the line. That's fine, but which error to send back if the device identifier is not supported by the API Implementation? example, an implementation is only able to perform an API with a phone number or IP but not networkAccessIdentifier.

**Possible evolution**
Agreed to use for this cas error 501 with code DEVICE_IDENTIFIER_TYPE_NOT_MANAGED

**Alternative solution**
use 500 (with code DEVICE_IDENTIFIER_TYPE_NOT_MANAGED) ?
Let this at the hand of each implem?

**Additional context**

### LINKED ISSUE #128: API Design : UNAUTHENTICATED 401
**ISSUE**
I am using IBM API Connect as a gateway where I don't have the control over internal authorization server. Due to this if any request fails due to the authorization issue, then it throws an error in the below format: 
{
    "httpCode": "401",
    "httpMessage": "Unauthorized",
    "moreInformation": "Invalid client id or secret."
}

Whereas expected CAMARA format is: 
{
  "status": 401,
  "code": "UNAUTHENTICATED",
  "message": "Request not authenticated due to missing, invalid, or expired credentials."
}

In this case the request gets rejected at the server level itself and it doesn't reach till my code to handle it. Due this I am unable to change the response format.

Currently, I am working with GSMA to close the certification for Sim Swap service but due to this particular use case it is not done.  
My suggestion is to ignore this use case for certification as I am sure other MNO's will also face this limitation.

### LINKED ISSUE #162: common.yaml unnecessarily list ErrorInfo properties as required.
**Problem description**

Assuming the language in API Design Guidelines is normative, CAMARA_common.yaml ErrorInfo format is incorrectly/unnencessarily requiring all the properties of ErrorInfo schema.
```
ErrorInfo:
   type: object
   required:
    - message
    - status
    - code
   properties:
    message:
     type: string
     description: A human readable description of what the event represent
    status:
     type: integer
     description: HTTP response status code
    code:
     type: string
     description: Friendly Code to describe the error
``` 

**Expected behavior**

Per DG, no field should be listed as required.

**Alternative solution**

Explicitly state which fields, if any, are required in Design Guideline and update, as necessary, the CAMARA_common.yaml to match definition.

**Additional context**

### LINKED ISSUE #172: Error Formatting with OpenID/OAuth 2.0 Standards
**Problem Description:** 
The current error formatting structure in the camera project diverges from the established standards of OpenID/OAuth 2.0. Although it includes fields like `status`, `code`, and `message`, as outlined in the provided `ErrorInfo`, it lacks key components such as `error`, `error_description`, and optionally `error_uri`, which are vital for effective error handling according to OpenID/OAuth 2.0 standards.

Here is the existing `ErrorInfo` structure:
```yaml
ErrorInfo:
  type: object
  required:
    - status
    - code
    - message
  properties:
    status:
      type: integer
      description: HTTP response status code
    code:
      type: string
      description: Code given to this error
    message:
      type: string
      description: Detailed error description
```

**Expected Behavior:** 
To ensure compliance with OpenID/OAuth 2.0 standards, the error formatting for the camera project should include the following fields:
- `error`: Represents the error code or identifier.
- `error_description` (optional): Provides a human-readable explanation of the error.
- `error_uri` (optional): Includes a URI that offers additional information about the error.

**Alternative Solution:** 
Alternatively, to align with OpenID/OAuth 2.0 standards, the camera project's error formatting can be updated as follows:

```yaml
error:
  type: object
  required:
    - error
    - error_description
  properties:
    error:
      type: string
      description: Error code or identifier.
    error_description:
      type: string
      description: Optional human-readable description of the error.
    error_uri:
      type: string
      description: Optional URI providing additional information about the error.
```

**Additional Context:** 
These suggested modifications closely adhere to the specifications outlined in the OpenID/OAuth 2.0 standard. For further clarity, please refer to the documentation provided in [RFC 6749 - The OAuth 2.0 Authorization Framework, Section 4.1.2.1](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.2.1).

### LINKED ISSUE #180: Revision of Chapter 6 Error Responses in API Design Guidelines
**Problem description**
There are a lot of questions and issues on Error Response format and content.
Currently the documentation lists the different error codes and cluster them into separate tables, depending on their nature: i) syntax exceptions, ii) service exceptions, and iii) server errors.
The tables do not contain HTTP status values which may introduce ambiguity.  There are also differences between the list and the examples in CAMARA_common.yaml

There are also questions on the Error Response format returned by OpenId Connect flows.

**Expected action**
Update Chapter 6 to provide clear requirements on Error Response content.
Decide what should be changed and aligned in [CAMARA_common.yaml](CAMARA_common.yaml).

Indicate when other standard formats can be used. 

**Additional context**
Resolution of Issue #127 can be extension/exemplification of Chapter 6 requirements.

### PR DESCRIPTION
**Pull request by PedroDiez**
#### What type of PR is this?

Add one of the following kinds:
* correction
* enhancement/feature
* documentation

#### What this PR does / why we need it:

This PR covers mainly 3 topics:
- Clarify error model properties are all mandatory. Issue #162
- Output for Issue #127
- Error Model Alignment #180 

Documentation Impacted:
- API Design Guidelines (sections 3.2 and section 6, new sections 6.1 and 6.2)
- CAMARA_common.yaml (aligment of message examples and ordering responses model by status)

#### Which issue(s) this PR fixes:

Fixes #127 
Fixes #128
Fixes #162
Fixes #172 
Fixes #180 

#### Special notes for reviewers:

Regarding API design guidelines, section 6.2 could also be documented in a separate file (.md for instance and refer from that section). I preferred to explicitly reflect to be easy to get the aim of the section. 

@rartych regarding Issue #180 in case you were thinking in additional points I can just take out (Fixes #180) and only reference in the (#### What this PR does / why we need it:) as partial fixing

#### Changelog input

```
 - Clarify Error properties are required
 - Error model for device Identifier
 - Added clarification about AuthN/AuthZ standardized flows follow their own Error Response format
 - Error model alignment

```

#### Additional documentation 

Not apply

### CODE REVIEW

File: documentation/API-design-guidelines.md [hunk 1]
```diff
@@ -631,7 +634,10 @@ An error representation must not be different from the representation of any res
 - A field "`status`", which can be identified in the response as a standard code from a list of Hypertext Transfer Protocol (HTTP) response status codes.
 - A unique error "`code`", which can be identified and traced for more details. It must be human readable; therefore, it must not be a numeric code. In turn, to achieve a better location of the error, you can reference the value or field that is causing it, and include it in the message. 
 - A detailed description of "`message`"
-  
+
+All these aforementioned fields are mandatory in Error Responses.
+`status` and `code` fields have normative nature, so as their use has to be normalize (see [Section 6.2](##62-normalization-use-of-error-responses)). On the other hand, `message` is informative and within this document an example is shown.
```
Comment (patrice-conil) [inline documentation/API-design-guidelines.md:639]: I think 'has to be normalized' or 'has to be standardized' would be better

**Review by patrice-conil — COMMENTED** [review]
Nice work ... one typo to fix but LGTM

**Comment by rartych** [comment]
To tackle #128 and #172 and similar questions I suggest to add paragraph like:

> When standardized authentication flows are used - cf. [10.2 Security Implementation](#102-security-implementation) the format and content of Error Response should follow relevant standards.

It can be in error format definition or particularly for 401 Status.

**Comment by jlurien** [comment]
I have linked another related discussion in QoD (https://github.com/camaraproject/QualityOnDemand/pull/297) to define the status and code error for situations when a request is rejected due to exceeding a business quota limit.  A new code QUOTA_EXCEEDED makes sense, but status could be either 429 or 403. The first being more related to rate limiting and the second a more generic "forbidden due to business logic". 

I will align the PR in QoD with the decision here. cc @PedroDiez

File: documentation/API-design-guidelines.md (hunk 1 above)
Comment (PedroDiez) [reply documentation/API-design-guidelines.md]: changed

**Comment by PedroDiez** [comment]
Actions covered in [commit 4611515ce7eedb10c961d14b1ccb3966dcdc447b](https://github.com/camaraproject/Commonalities/tree/4611515ce7eedb10c961d14b1ccb3966dcdc447b)
- Suggestion by Patrice
- Suggestions by Rafal to unify info and add clarification for Issues#128 and Issue#172
- Added `QUOTA_EXCEEDED` code, based on topic raised by Jose. Proposed to be 429 status in the sense it is focused in a business quota limit that has been overpassed (e.g. number of N QoD minutes or sessions per period). Also fine to be 403 in case seen more suitable.

**Comment by PedroDiez** [comment]
Actions covered in [commit 082f5e52ddec8920432fe12d97cbce62e4cb5fb3](https://github.com/camaraproject/Commonalities/tree/082f5e52ddec8920432fe12d97cbce62e4cb5fb3)

- Grouping in responses section all the ones that own the same status due to an API that responds to a status must reference a single responses object (it cannot indicate that it is an anyOf and show N $refs).

File: documentation/API-design-guidelines.md [hunk 2]
```diff
@@ -651,44 +658,120 @@ The essential requirements to consider would be:
...
+| 405 | `METHOD_NOT_ALLOWED` | The requested method is not allowed/supported on the target resource. | Invalid HTTP verb used with a given endpoint |
+| 406 | `NOT_ACCEPTABLE` | The server cannot produce a response matching the content requested by the client through `Accept-*` headers. | API Server does not accept the media type (`Accept-*` header) indicated by API client |
+| 410 | `GONE` | Access to the target resource is no longer available. | Use in notifications flow to allow API Consumer to indicate that its callback is no longer available |
+| 412 | `FAILED_PRECONDITION` | Request cannot be executed in the current system state. | Indication by the API Server that the request cannot be processed in current system state |
+| 415 | `UNSUPPORTED_MEDIA_TYPE` | The server refuses to accept the request because the payload format is in an unsupported format. | Payload format of the request is in an unsupported format by the Server. Should not happen |
+| 500 | `DATA_LOSS` | Unrecoverable data loss or data corruption. | Some data is lost or corrupted when the server is processing the request so as it cannot manage it properly. It may point to a potential problem at DD.BB. level |
```
Comment (rartych) [inline documentation/API-design-guidelines.md:711]: Do we need this error specified? 
If so, then `DD.BB. level` is not clear for me.

Comment (PedroDiez) [reply documentation/API-design-guidelines.md:711]: In the description i was referring that when this happens one of the reasons can be at DD.BB. level (problem in DD.BB. system providing support to API BE). It is fine to me to rephrase to:

_Some data is lost or corrupted when the server is processing the request so as it cannot manage it properly_


On the other hand, Currently it is not used in any API. So, we could take out so far 
WDYT? @rartych, @patrice-conil

File: documentation/API-design-guidelines.md [hunk 3]
```diff
@@ -651,44 +658,120 @@ The essential requirements to consider would be:
...
 
-| **Error code** | Description |
-|-----|-----|
-|`INVALID_ARGUMENT` |  A specified resource duplicate entry found. |
-|`CONFLICT` |  A specified resource duplicate entry found. |
-|`OUT_OF_RANGE`| Client specified an invalid range. |
-|`PERMISSION_DENIED`|  Client does not have sufficient permissions to perform this action. |
-|`ABORTED` |  Concurrency conflict.|
-|`ALREADY_EXISTS` | The resource that a client tried to create already exists. |
+| **Error status** | **Error code** | **Message** | **scope/description** |
```
Comment (rartych) [inline documentation/API-design-guidelines.md:674]:
```suggestion
| **Error status** | **Error code** | **Message example** | **Scope/description** |
```

File: documentation/API-design-guidelines.md [hunk 4]
```diff
@@ -651,44 +658,120 @@ The essential requirements to consider would be:
...
 
 <font size="3"><span style="color: blue"> Service Exceptions </span></font>
 
-| **Error code** | Description |
-|-----|-----|
-|`UNAUTHENTICATED` |  Request not authenticated due to missing, invalid, or expired credentials.|
-|`NOT_FOUND`|  The specified resource is not found. |
-|`TOO_MANY_REQUESTS`| Either out of resource quota or reaching rate limiting. |
-|`AUTHENTICATION_REQUIRED`|  New authentication is required. |
+| **Error status** | **Error code** | **Message** | **scope/description** |
```
Comment (rartych) [inline documentation/API-design-guidelines.md:688]:
```suggestion
| **Error status** | **Error code** | **Message example** | **Scope/description** |
```

File: documentation/API-design-guidelines.md [hunk 5]
```diff
@@ -651,44 +658,120 @@ The essential requirements to consider would be:
...
-|`BAD_GATEWAY`| An upstream internal service cannot be reached. |
-|`UNAVAILABLE`| Request timeout exceeded. |
-|`TIMEOUT`| Request timeout exceeded. |
-|`NOT_IMPLEMENTED`| This functionality is not implemented yet |
-|`METHOD_NOT_ALLOWED`|	The requested method is not allowed/supported on the target resource. |
-|`NOT_ACCEPTABLE` | The server cannot produce a response matching the content requested by the client through `Accept-*` headers. |
-|`UNSUPPORTED_MEDIA_TYPE`	| The server refuses to accept the request because the payload format is in an unsupported format. |
-
-> _NOTE: When no login has been performed or no authentication has been assigned, a non-descriptive generic error will always be returned in all cases, a `UNAUTHENTICATED` 401 “Request not authenticated due to missing, invalid, or expired credentials.” is returned, whatever the reason._
+| **Error status** | **Error code** | **Message** | **scope/description** |
```
Comment (rartych) [inline documentation/API-design-guidelines.md:704]:
```suggestion
| **Error status** | **Error code** | **Message example** | **Scope/description** |
```

File: documentation/API-design-guidelines.md [hunk 6]
```diff
@@ -651,44 +658,120 @@ The essential requirements to consider would be:
...
+
+> _NOTE 2: A {{SPECIFIC_CODE}}, unless it may have traversal scope (i.e. re-usable among different APIs), SHALL follow this scheme for a specific API: {{API_NAME}}.{{SPECIFIC_CODE}}
+
+### 6.2 Error Responses - Device Object
+
+This section is focused in the guidelines about error responses around the concept of `device` object.
+
+Following table compiles the guidelines to be adopted:
+
+| Case # | validation_description | status_value | code_value | message_example |
```
Comment (rartych) [inline documentation/API-design-guidelines.md:728]: I propose to align table headers with the table above:
```suggestion
| **Case #** | **Description** | **Error status** | **Error code**  | **Message example** |
```
| **Error status** | **Error code** | **Message example** | **Scope/description** |

File: documentation/API-design-guidelines.md [hunk 7]
```diff
@@ -651,44 +658,120 @@ The essential requirements to consider would be:
...
+| 0 | The request body does not comply with the schema | 400 | INVALID_ARGUMENT | device does not comply with the schema. |
+| 1 | None of the provided device identifiers is supported by the implementation | 422 | UNSUPPORTED_DEVICE_IDENTIFIERS | phoneNumber is required. |
+| 2 | Some identifier cannot be matched to a device | 404 | DEVICE_NOT_FOUND | Device identifier not found. |  
+| 3 | Device identifiers mismatch | 422 | DEVICE_IDENTIFIERS_MISMATCH | Provided device identifiers are not consistent. |
+| 4 | Invalid access token context | 403 | INVALID_TOKEN_CONTEXT | Device identifiers are not consistent with access token. |
+| 5 | Service not applicable to the device | 422 | DEVICE_NOT_APPLICABLE | The service is not available for the provided device. |
+
+#### Templates
+
+##### response template
```
Comment (rartych) [inline documentation/API-design-guidelines.md:739]:
```suggestion
##### Response template
```

File: documentation/API-design-guidelines.md [hunk 8]
```diff
@@ -651,44 +658,120 @@ The essential requirements to consider would be:
...
+
+A response will group all examples for same operation and status code
+Schema is common for all errors
+
+```
```
Comment (rartych) [inline documentation/API-design-guidelines.md:744]: Better rendering:
```suggestion
```yaml
```

File: documentation/API-design-guidelines.md [hunk 9]
```diff
@@ -651,44 +658,120 @@ The essential requirements to consider would be:
...
+        $ref: ""#/components/examples/{{case_1}}"
+      {{case_2}}:
+        $ref: ""#/components/examples/{{case_2}}"      
+```
+
+##### example template
+
+One case will be needed per row in the table above, following this template:
+
+```
```
Comment (rartych) [inline documentation/API-design-guidelines.md:764]: Better rendering:
```suggestion
```yaml
```

**Review by bigludo7 — APPROVED** [review]

**Comment by PedroDiez** [comment]
05/JUN: Addressed first round of reviews


Pending point: 
- Exception 500 DATA_LOSS. Should we take out so far? (please see above comment @rartych, @patrice-conil )

**Review by patrice-conil — COMMENTED** [review]
I support removing DATA_LOSS because I don't see any use cases using it

File: documentation/API-design-guidelines.md (hunk 2 above)
Comment (PedroDiez) [reply documentation/API-design-guidelines.md]: removed `DATA_LOSS` error code

**Review by shilpa-padgaonkar — APPROVED** [review]
/LGTM

**Review by rartych — APPROVED** [review]

**Pull request merged**